lib/task.ml

```diff
@@ -203,29 +203,36 @@ let parallel_for ?(chunk_size=0) ~start ~finish ~body pool =
 
 let parallel_scan pool op elements =
   let pd = get_pool_data pool in
+  let n = Array.length elements in
+  let p = min (n - 1) ((Array.length pd.domains) + 1) in
+  let prefix_s = Array.copy elements in
   let scan_part op elements prefix_sum start finish =
     assert (Array.length elements > (finish - start));
     for i = (start + 1) to finish do
       prefix_sum.(i) <- op prefix_sum.(i - 1) elements.(i)
     done
   in
+  if p = 1 then begin
+    (* Do a sequential scan when number of domains is 1, array's length is 1
+    *)
+    scan_part op elements prefix_s 0 (n - 1);
+    prefix_s
+  end
+  else begin
   let add_offset op prefix_sum offset start finish =
     assert (Array.length prefix_sum > (finish - start));
     for i = start to finish do
       prefix_sum.(i) <- op offset prefix_sum.(i)
     done
   in
-  let n = Array.length elements in
-  let p = (Array.length pd.domains) + 1 in
-  let prefix_s = Array.copy elements in
 
   parallel_for pool ~chunk_size:1 ~start:0 ~finish:(p - 1)
   ~body:(fun i ->
     let s = (i * n) / (p ) in
     let e = (i + 1) * n / (p ) - 1 in
     scan_part op elements prefix_s s e);
 
-  if (p > 2) then begin
+  if (p > 1) then begin
```

Comment:
If I am reading this correctly
- first this now solves the `p=1` case in the `then`-branch and
- then handles cases `p=-1,0 + 2,3,4,...` in the `else`-branch with a sub-case test for `p>1` (`p=2,3,4,...`)

Could this be simplified to:
- solve the `p<2` cases in the then-branch and
- then handles cases `p=2,3,4,...` in the else-branch?

Structurally I think it would read simpler. 
IIUC, `p=-1,0,1` happens when elements has 0,1,2 elements or there's only 1 domain.
In both these situations, a sequential scan should be OK.

Reply:
That's a good idea! Done.